File: valo_tui/screens/brackets.py
"""[b] brackets — ASCII double-elimination trees.

The bracket is reconstructed in :mod:`valo_tui.data.bracket`; this module lays
it out on a character grid (midpoint positioning, box-drawing connectors) and
renders it as a selectable, navigable widget."""

from __future__ import annotations

from rich.markup import escape

from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Horizontal, VerticalScroll
from textual.reactive import reactive
from textual.screen import Screen
from textual.widgets import Footer, Label, ListItem, ListView, Static

from ..data import cache
from ..data.bracket import Bracket, BracketMatch
from ..data.models import EventCard

# Grid geometry.
BOX_W = 15          # team box: name (11) + score (4)
CONN = 7            # connector zone width between columns
COLW = BOX_W + CONN

ACCENT = "#e87a5d"
TEXT = "#c8d8e8"
MUTED = "#6e8aa6"
DIM = "#2f4a5f"
SEL_BG = "#16344a"


# ── layout ───────────────────────────────────────────────────
def _layout_section(columns) -> tuple[dict[int, tuple[int, int, int]], int]:
    """Assign (top, mid, bot) rows to each match; return positions + height."""
    pos: dict[int, tuple[int, int, int]] = {}
    height = 0
    prev = None
    for col in columns:
        winners_prev: dict[str, BracketMatch] = {}
        if prev:
            for pm in prev.matches:
                if pm.winner_name:
                    winners_prev[pm.winner_name] = pm

        guesses: dict[int, int | None] = {}
        for m in col.matches:
            anchors = [
                pos[f.match_id][1]
                for f in (winners_prev.get(m.top.name), winners_prev.get(m.bottom.name))
                if f is not None and f.match_id in pos
            ]
            guesses[m.match_id] = sum(anchors) // len(anchors) if anchors else None

        order = sorted(
            col.matches,
            key=lambda m: (
                guesses[m.match_id] is None,
                guesses[m.match_id] or 0,
                col.matches.index(m),
            ),
        )
        cur = 0
        for m in order:
            g = guesses[m.match_id]
            mid = max(cur + 1, g if g is not None else cur + 1)
            pos[m.match_id] = (mid - 1, mid, mid + 1)
            cur = mid + 3
            height = max(height, mid + 2)
        prev = col
    return pos, height


class _Grid:
    """A character grid with a parallel style grid, emitted as Rich Text."""

    def __init__(self, w: int, h: int) -> None:
        self.w, self.h = w, h
        self.ch = [[" "] * w for _ in range(h)]
        self.st = [[""] * w for _ in range(h)]

    def put(self, y: int, x: int, text: str, style: str = "") -> None:
        if not (0 <= y < self.h):
            return
        for i, c in enumerate(text):
            if 0 <= x + i < self.w:
                self.ch[y][x + i] = c
                self.st[y][x + i] = style

    def to_markup(self) -> str:
        lines = []
        for y in range(self.h):
            parts, run, run_style = [], "", ""
            for x in range(self.w):
                c, s = self.ch[y][x], self.st[y][x]
                if s != run_style:
                    parts.append(_wrap(run, run_style))
                    run, run_style = c, s
                else:
                    run += c
            parts.append(_wrap(run, run_style))
            lines.append("".join(parts).rstrip())
        return "\n".join(lines)


def _wrap(text: str, style: str) -> str:
    if not text:
        return ""
    safe = escape(text)
    return f"[{style}]{safe}[/]" if style else safe


def _box(slot, selected: bool) -> tuple[str, str]:
    name = (slot.name or "TBD")[:11]
    score = "·" if slot.score is None else str(slot.score)
    text = f"{name:<11}{score:>4}"
    if slot.winner:
        style = f"bold {ACCENT}"
    else:
        style = MUTED
    if selected:
        style = f"bold {TEXT} on {SEL_BG}"
    return text, style


def render_bracket(bracket: Bracket, selected_id: int | None) -> str:
    grid = build_grid(bracket, selected_id)
    return grid.to_markup()


def build_grid(bracket: Bracket, selected_id: int | None) -> "_Grid":
    sections = bracket.sections
    if not sections:
        g = _Grid(16, 1)
        g.put(0, 0, "no bracket data", MUTED)
        return g

    # Pre-compute per-section layout + dimensions.
    layouts = [_layout_section(s.columns) for s in sections]
    width = max((len(s.columns) * COLW for s in sections), default=COLW)
    total_h = sum(h for _, h in layouts) + 3 * len(sections)
    grid = _Grid(width + 2, total_h + 1)

    row_off = 0
    for sec, (pos, sec_h) in zip(sections, layouts):
        grid.put(row_off, 0, sec.name.upper(), f"bold {ACCENT}")
        base = row_off + 2
        prev = None
        for ci, col in enumerate(sec.columns):
            x0 = ci * COLW
            # column header
            grid.put(row_off + 1, x0, col.title[: BOX_W + 2], MUTED)
            winners_prev: dict[str, BracketMatch] = {}
            if prev:
                for pm in prev.matches:
                    if pm.winner_name:
                        winners_prev[pm.winner_name] = pm
            for m in col.matches:
                top, mid, bot = pos[m.match_id]
                sel = m.match_id == selected_id
                t_txt, t_st = _box(m.top, sel)
                b_txt, b_st = _box(m.bottom, sel)
                grid.put(base + top, x0, t_txt, t_st)
                grid.put(base + bot, x0, b_txt, b_st)
                if ci > 0:
                    _draw_conn(grid, base, x0, ci, m, mid, winners_prev, pos)
            prev = col
        row_off += sec_h + 3

    return grid


def _draw_conn(grid, base, x0, ci, m, mid, winners_prev, pos) -> None:
    feeders = [
        winners_prev.get(m.top.name),
        winners_prev.get(m.bottom.name),
    ]
    rows = [pos[f.match_id][1] for f in feeders if f is not None and f.match_id in pos]
    if not rows:
        return
    chan = x0 - 4
    prev_end = (ci - 1) * COLW + BOX_W

    # single feeder, same row -> straight line
    if len(rows) == 1 and rows[0] == mid:
        grid.put(base + mid, prev_end, "─" * (x0 - prev_end), DIM)
        return

    lo, hi = min(rows + [mid]), max(rows + [mid])
    for y in range(lo, hi + 1):
        if grid.ch[base + y][chan] == " ":
            grid.put(base + y, chan, "│", DIM)
    for fy in rows:
        grid.put(base + fy, prev_end, "─" * (chan - prev_end), DIM)
        corner = "┐" if fy < mid else ("┘" if fy > mid else "├")
        grid.put(base + fy, chan, corner, DIM)
    grid.put(base + mid, chan, "├", DIM)
    grid.put(base + mid, chan + 1, "─" * (x0 - chan - 1), DIM)


# ── widget ───────────────────────────────────────────────────
class BracketWidget(Static):
    """Renders a bracket and tracks a selected match for navigation."""

    BINDINGS = [
        Binding("j,down", "move(0, 1)", "Down", show=False),
        Binding("k,up", "move(0, -1)", "Up", show=False),
        Binding("l,right", "move(1, 0)", "Right", show=False),
        Binding("h,left", "move(-1, 0)", "Left", show=False),
        Binding("enter", "open", "Open match", show=False),
    ]
    can_focus = True
    sel_col: reactive[int] = reactive(0)
    sel_row: reactive[int] = reactive(0)

    def __init__(self, bracket: Bracket, **kwargs) -> None:
        self.bracket = bracket
        # Flatten all columns across sections for h/l navigation.
        self.columns = [c for s in bracket.sections for c in s.columns]
        first = self.columns[0].matches[0].match_id if self.columns and self.columns[0].matches else None
        grid = build_grid(bracket, first)
        self._w, self._h = grid.w, grid.h
        super().__init__(grid.to_markup(), **kwargs)
        # Explicit size avoids Textual's auto-width content measurement.
        self.styles.width = self._w
        self.styles.height = self._h

    def _selected_match(self) -> BracketMatch | None:
        if not self.columns:
            return None
        col = self.columns[self.sel_col]
        if not col.matches:
            return None
        return col.matches[self.sel_row]

    def action_move(self, dc: int, dr: int) -> None:
        if not self.columns:
            return
        if dc:
            self.sel_col = max(0, min(len(self.columns) - 1, self.sel_col + dc))
            self.sel_row = min(self.sel_row, len(self.columns[self.sel_col].matches) - 1)
        if dr:
            n = len(self.columns[self.sel_col].matches)
            self.sel_row = max(0, min(n - 1, self.sel_row + dr))
        self._redraw()

    def action_open(self) -> None:
        m = self._selected_match()
        if m and m.match_id:
            from .match_detail import MatchDetailScreen

            self.app.push_screen(MatchDetailScreen(m.match_id))

    def _redraw(self) -> None:
        m = self._selected_match()
        self.update(render_bracket(self.bracket, m.match_id if m else None))


# ── screen ───────────────────────────────────────────────────
class BracketsScreen(Screen):
    BINDINGS = [Binding("escape,q", "app.pop_screen", "Back")]

    def __init__(self, event_id: int | None = None) -> None:
        super().__init__()
        self._event_id = event_id

    def compose(self) -> ComposeResult:
        with Horizontal():
            with VerticalScroll(id="bracket-events"):
                yield Label("events", classes="page-title")
                yield ListView(id="event-list")
            with VerticalScroll(id="bracket-body"):
                yield Label("select an event  ·  enter", id="bracket-hint", classes="hint")
        yield Footer()

    def on_mount(self) -> None:
        events = cache.active_events()
        lv = self.query_one("#event-list", ListView)
        self._events: list[EventCard] = events
        for e in events:
            lv.append(ListItem(Label(e.name[:30]), id=f"ev-{e.id}"))
        # Only fetch when the user explicitly picks an event (or one was passed
        # in). We never auto-ping the API just because the screen opened.
        if self._event_id is not None:
            self._load(self._event_id)

    def on_list_view_selected(self, event: ListView.Selected) -> None:
        if event.item.id and event.item.id.startswith("ev-"):
            self._load(int(event.item.id[3:]))

    def _load(self, event_id: int) -> None:
        body = self.query_one("#bracket-body", VerticalScroll)
        body.remove_children()
        bracket = cache.bracket(event_id)
        if not bracket.has_data:
            body.mount(Label("no playoff bracket for this event yet", classes="hint"))
            return
        widget = BracketWidget(bracket)
        body.mount(widget)
        self.call_after_refresh(widget.focus)

File: internal/screens/brackets.go
package screens

import (
	"fmt"
	"sort"
	"strings"

	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"

	"github.com/jashkarangiya/valo-tui/internal/data"
	"github.com/jashkarangiya/valo-tui/internal/styles"
)

// Grid geometry (matches screens/brackets.py).
const (
	boxW = 15 // team box: name(11) + score(4)
	conn = 7  // connector zone between columns
	colW = boxW + conn
)

// Cell style tags for the bracket grid.
type cellTag int

const (
	tagPlain cellTag = iota
	tagWinner
	tagMuted
	tagDim
	tagSelected
	tagSection
)

var bracketStyles = map[cellTag]lipgloss.Style{
	tagPlain:    lipgloss.NewStyle().Foreground(styles.Text),
	tagWinner:   lipgloss.NewStyle().Foreground(styles.Accent).Bold(true),
	tagMuted:    lipgloss.NewStyle().Foreground(styles.Muted),
	tagDim:      lipgloss.NewStyle().Foreground(styles.Rule),
	tagSelected: lipgloss.NewStyle().Foreground(styles.Text).Background(styles.SelBg).Bold(true),
	tagSection:  lipgloss.NewStyle().Foreground(styles.Accent).Bold(true),
}

// grid is a character grid with a parallel style-tag grid.
type grid struct {
	w, h int
	ch   [][]rune
	tag  [][]cellTag
}

func newGrid(w, h int) *grid {
	if w < 1 {
		w = 1
	}
	if h < 1 {
		h = 1
	}
	ch := make([][]rune, h)
	tg := make([][]cellTag, h)
	for y := 0; y < h; y++ {
		ch[y] = make([]rune, w)
		tg[y] = make([]cellTag, w)
		for x := 0; x < w; x++ {
			ch[y][x] = ' '
		}
	}
	return &grid{w: w, h: h, ch: ch, tag: tg}
}

func (g *grid) put(y, x int, s string, tag cellTag) {
	if y < 0 || y >= g.h {
		return
	}
	for i, c := range []rune(s) {
		if x+i >= 0 && x+i < g.w {
			g.ch[y][x+i] = c
			g.tag[y][x+i] = tag
		}
	}
}

func (g *grid) String() string {
	lines := make([]string, g.h)
	for y := 0; y < g.h; y++ {
		var b strings.Builder
		run, runTag := strings.Builder{}, g.tag[y][0]
		flush := func() {
			if run.Len() > 0 {
				b.WriteString(bracketStyles[runTag].Render(run.String()))
				run.Reset()
			}
		}
		for x := 0; x < g.w; x++ {
			t := g.tag[y][x]
			if t != runTag {
				flush()
				runTag = t
			}
			run.WriteRune(g.ch[y][x])
		}
		flush()
		lines[y] = strings.TrimRight(b.String(), " ")
	}
	return strings.Join(lines, "\n")
}

type rowPos struct{ top, mid, bot int }

// layoutSection assigns (top,mid,bot) rows to each match; returns positions and
// the section height. Ported from _layout_section.
func layoutSection(columns []data.BracketColumn) (map[int]rowPos, int) {
	pos := map[int]rowPos{}
	height := 0
	var prev *data.BracketColumn
	for ci := range columns {
		col := columns[ci]
		winnersPrev := map[string]data.BracketMatch{}
		if prev != nil {
			for _, pm := range prev.Matches {
				if w := pm.WinnerName(); w != "" {
					winnersPrev[w] = pm
				}
			}
		}

		guesses := map[int]int{} // match_id → guessed mid
		hasGuess := map[int]bool{}
		for _, m := range col.Matches {
			var anchors []int
			for _, name := range []string{m.Top.Name, m.Bottom.Name} {
				if f, ok := winnersPrev[name]; ok {
					if p, ok := pos[f.MatchID]; ok {
						anchors = append(anchors, p.mid)
					}
				}
			}
			if len(anchors) > 0 {
				sum := 0
				for _, a := range anchors {
					sum += a
				}
				guesses[m.MatchID] = sum / len(anchors)
				hasGuess[m.MatchID] = true
			}
		}

		order := append([]data.BracketMatch(nil), col.Matches...)
		idxOf := map[int]int{}
		for i, m := range col.Matches {
			idxOf[m.MatchID] = i
		}
		sort.SliceStable(order, func(a, b int) bool {
			ga, gb := hasGuess[order[a].MatchID], hasGuess[order[b].MatchID]
			if ga != gb {
				return ga // those with a guess first
			}
			if guesses[order[a].MatchID] != guesses[order[b].MatchID] {
				return guesses[order[a].MatchID] < guesses[order[b].MatchID]
			}
			return idxOf[order[a].MatchID] < idxOf[order[b].MatchID]
		})

		cur := 0
		for _, m := range order {
			mid := cur + 1
			if hasGuess[m.MatchID] && guesses[m.MatchID] > mid {
				mid = guesses[m.MatchID]
			}
			pos[m.MatchID] = rowPos{mid - 1, mid, mid + 1}
			cur = mid + 3
			if mid+2 > height {
				height = mid + 2
			}
		}
		prev = &columns[ci]
	}
	return pos, height
}

func bracketBox(slot data.BracketSlot, selected bool) (string, cellTag) {
	name := clipRunes(slot.Name, 11)
	if name == "" {
		name = "TBD"
	}
	score := "·"
	if slot.Score != nil {
		score = fmt.Sprint(*slot.Score)
	}
	txt := fmt.Sprintf("%-11s%4s", name, score)
	tag := tagMuted
	if slot.Winner {
		tag = tagWinner
	}
	if selected {
		tag = tagSelected
	}
	return txt, tag
}

func buildBracketGrid(b data.Bracket, selectedID int) *grid {
	if len(b.Sections) == 0 {
		g := newGrid(16, 1)
		g.put(0, 0, "no bracket data", tagMuted)
		return g
	}
	type secLayout struct {
		pos    map[int]rowPos
		height int
	}
	layouts := make([]secLayout, len(b.Sections))
	width := colW
	totalH := 0
	for i, s := range b.Sections {
		pos, h := layoutSection(s.Columns)
		layouts[i] = secLayout{pos, h}
		if w := len(s.Columns) * colW; w > width {
			width = w
		}
		totalH += h
	}
	totalH += 3 * len(b.Sections)
	g := newGrid(width+2, totalH+1)

	rowOff := 0
	for si, sec := range b.Sections {
		lay := layouts[si]
		g.put(rowOff, 0, strings.ToUpper(sec.Name), tagSection)
		base := rowOff + 2
		var prev *data.BracketColumn
		for ci := range sec.Columns {
			col := sec.Columns[ci]
			x0 := ci * colW
			g.put(rowOff+1, x0, clipRunes(col.Title, boxW+2), tagMuted)
			winnersPrev := map[string]data.BracketMatch{}
			if prev != nil {
				for _, pm := range prev.Matches {
					if w := pm.WinnerName(); w != "" {
						winnersPrev[w] = pm
					}
				}
			}
			for _, m := range col.Matches {
				p := lay.pos[m.MatchID]
				sel := m.MatchID == selectedID
				tTxt, tTag := bracketBox(m.Top, sel)
				bTxt, bTag := bracketBox(m.Bottom, sel)
				g.put(base+p.top, x0, tTxt, tTag)
				g.put(base+p.bot, x0, bTxt, bTag)
				if ci > 0 {
					drawConn(g, base, x0, ci, m, p.mid, winnersPrev, lay.pos)
				}
			}
			prev = &sec.Columns[ci]
		}
		rowOff += lay.height + 3
	}
	return g
}

func drawConn(g *grid, base, x0, ci int, m data.BracketMatch, mid int, winnersPrev map[string]data.BracketMatch, pos map[int]rowPos) {
	var rows []int
	for _, name := range []string{m.Top.Name, m.Bottom.Name} {
		if f, ok := winnersPrev[name]; ok {
			if p, ok := pos[f.MatchID]; ok {
				rows = append(rows, p.mid)
			}
		}
	}
	if len(rows) == 0 {
		return
	}
	chan_ := x0 - 4
	prevEnd := (ci-1)*colW + boxW

	if len(rows) == 1 && rows[0] == mid {
		g.put(base+mid, prevEnd, strings.Repeat("─", x0-prevEnd), tagDim)
		return
	}
	lo, hi := mid, mid
	for _, r := range rows {
		if r < lo {
			lo = r
		}
		if r > hi {
			hi = r
		}
	}
	for y := lo; y <= hi; y++ {
		if base+y >= 0 && base+y < g.h && g.ch[base+y][chan_] == ' ' {
			g.put(base+y, chan_, "│", tagDim)
		}
	}
	for _, fy := range rows {
		g.put(base+fy, prevEnd, strings.Repeat("─", chan_-prevEnd), tagDim)
		corner := "├"
		if fy < mid {
			corner = "┐"
		} else if fy > mid {
			corner = "┘"
		}
		g.put(base+fy, chan_, corner, tagDim)
	}
	g.put(base+mid, chan_, "├", tagDim)
	g.put(base+mid, chan_+1, strings.Repeat("─", x0-chan_-1), tagDim)
}

// Bracket is the [b] event sub-page: the playoff tree, navigable with h/j/k/l.
type Bracket struct {
	w, h    int
	bracket data.Bracket
	columns []data.BracketColumn // flattened for h/l navigation
	selCol  int
	selRow  int
	hasEvt  bool
}

func NewBracket(w, h int) Bracket { return Bracket{w: w, h: h} }

func (s *Bracket) SetSize(w, h int) { s.w, s.h = w, h }

func (s *Bracket) Load(eventID int) {
	s.hasEvt = eventID != 0
	if !s.hasEvt {
		s.bracket = data.Bracket{}
		s.columns = nil
		s.selCol, s.selRow = 0, 0
		return
	}
	s.bracket = data.BracketFor(eventID)
	s.columns = nil
	for _, sec := range s.bracket.Sections {
		s.columns = append(s.columns, sec.Columns...)
	}
	// Preserve the cursor across periodic reloads, clamped to the new bounds.
	s.selCol = clampIndex(s.selCol, len(s.columns))
	if s.selCol < len(s.columns) {
		s.selRow = clampIndex(s.selRow, len(s.columns[s.selCol].Matches))
	}
}

// clampIndex keeps i within [0, n-1], returning 0 for an empty collection.
func clampIndex(i, n int) int {
	if i < 0 || n == 0 {
		return 0
	}
	if i > n-1 {
		return n - 1
	}
	return i
}

func (s *Bracket) selectedMatch() (data.BracketMatch, bool) {
	if s.selCol >= len(s.columns) {
		return data.BracketMatch{}, false
	}
	col := s.columns[s.selCol]
	if s.selRow >= len(col.Matches) {
		return data.BracketMatch{}, false
	}
	return col.Matches[s.selRow], true
}

// SelectedMatchID is the match the cursor is on (0 if none).
func (s Bracket) SelectedMatchID() int {
	if m, ok := s.selectedMatch(); ok {
		return m.MatchID
	}
	return 0
}

func (s *Bracket) Focus() {}
func (s *Bracket) Blur()  {}

func (s Bracket) Update(msg tea.Msg) (Bracket, tea.Cmd) {
	k, ok := msg.(tea.KeyPressMsg)
	if !ok || len(s.columns) == 0 {
		return s, nil
	}
	switch k.String() {
	case "l", "right":
		if s.selCol < len(s.columns)-1 {
			s.selCol++
		}
		if n := len(s.columns[s.selCol].Matches); s.selRow > n-1 {
			s.selRow = n - 1
		}
	case "h", "left":
		if s.selCol > 0 {
			s.selCol--
		}
		if n := len(s.columns[s.selCol].Matches); s.selRow > n-1 {
			s.selRow = n - 1
		}
	case "j", "down":
		if n := len(s.columns[s.selCol].Matches); s.selRow < n-1 {
			s.selRow++
		}
	case "k", "up":
		if s.selRow > 0 {
			s.selRow--
		}
	}
	return s, nil
}

func (s Bracket) View() string {
	header := title("bracket") + "\n" + hint("h/j/k/l move · enter → scoreboards") + "\n\n"
	if !s.hasEvt {
		return header + muted("select an event first")
	}
	if !s.bracket.HasData() {
		return header + muted("no playoff bracket for this event yet")
	}
	return header + buildBracketGrid(s.bracket, s.SelectedMatchID()).String()
}
